service/resource/endpoint/delete.go

```diff
@@ -53,26 +23,35 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteState inter
 }
 
 func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
-	delete, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
+	endpoint, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sEndpoint, err := constructK8sEndpoint(endpoint, r.k8sClient)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
 	patch := framework.NewPatch()
-	patch.SetDeleteChange(delete)
+	if len(endpoint.IPs) == 0 {
```

Comment:
No IPs means that the Endpoint would be empty, therefore we delete it then. Otherwise we update the endpoint with one missing IP.

Reply:
IMO the `New*Patch` methods should be as similar as possible and not contain any logic but gathering changes from the private `new*Change` methods. Pawel and I agreed on this recently.

Reply:
Okay I missed this, I thought the `new*Change` functions were just for the transition to the operatorkit update. I'll add them back.

Reply:
The way I understand it now is that `New*Patch` is just a simple dispatcher. It dispatches `new*Change` to `Apply*Change`. In most cases `New*Patch` is identical. We agreed that `Apply*Change` should be as dump as possible. Best case is reduced to single API call + logging + `if` statement checking if the `change` is set to make the logging possible. 

`New*Patch` example:
https://github.com/giantswarm/azure-operator/blob/0250ba9bc7deb8b090fb54f0ba4cf08accf7b939/service/resource/resourcegroup/resource.go#L129-L139

`Apply*Change` example:
https://github.com/giantswarm/azure-operator/blob/0250ba9bc7deb8b090fb54f0ba4cf08accf7b939/service/resource/resourcegroup/resource.go#L161-L197

Reply:
I tried to simplify both functions as much as I could. There are not only simple sanity checks and the decision what patch should be done in there. I am not certain if this is simple enough / the way you want it to be.